arm_robots/src/arm_robots/cartesian.py

```diff
@@ -0,0 +1,221 @@
+        self._start_violation = 0
+
+        self.active_arm = ArmSide.LEFT
+        # ROS subscribers
+        self.motion_status_listeners = motion_status_listeners
+
+        # ROS publishers
+        self.motion_command_publisher = motion_command_publisher
+
+        # TODO should probably move control loop onto separate thread
+
+    def set_active_arm(self, active_arm: ArmSide):
+        self.active_arm = active_arm
+
+    def reset(self):
+        self.abort_goal()
+
+    def abort_goal(self):
+        self.target_pose = None
+        self._intermediate_target = None
+        self._dists_to_goal = []
+        self._goal_start_time = None
+        self._init_goal_dist = None
+        self._check_joint_limits = True
+        self._start_violation = 0
+
+    def current_pose_in_world_frame(self, arm):
+        current_pose = self.motion_status_listeners[arm].get().measured_cartesian_pose
+        if current_pose is None:
+            return None
+        tf_current_pose = PoseStamped()
+        tf_current_pose.pose = current_pose
+        tf_current_pose.header.frame_id = self.sensor_frames[arm]
+        return self.tf_buffer.transform(tf_current_pose, self.world_frame)
+
+    def set_relative_goal_2d(self, dx, dy, target_z=None, target_orientation=None):
+        cp = self.current_pose_in_world_frame(self.active_arm)
+        if cp is None:
+            rospy.logwarn("Trying to set relative goal when we do not have current pose")
+            return False
+
+        target_pose = copy.deepcopy(cp)
+        target_pose.pose.position.x += dx
+        target_pose.pose.position.y += dy
+        if target_z is not None:
+            target_pose.pose.position.z = target_z
+        if target_orientation is not None:
+            target_pose.pose.orientation = target_orientation
+
+        self.set_target_pose(target_pose, current_pose=cp)
+
+        rospy.logdebug("Current {}".format(cp.pose))
+        return True
+
+    def set_target_pose(self, target_pose, current_pose=None):
+        if current_pose is None:
+            current_pose = self.current_pose_in_world_frame(self.active_arm)
+        self.target_pose = copy.deepcopy(target_pose)
+        self._init_goal_dist = pose_distance(current_pose.pose, self.target_pose.pose)
+        self._start_violation = self.joint_boundary_violation_amount()
+        self._goal_start_time = rospy.get_time()
+        self.timed_out = False
+        self.reached_joint_limit = False
+        rospy.logdebug("Target {}".format(self.target_pose.pose))
+
+    def joint_boundary_violation_amount(self):
+        q = self.motion_status_listeners[self.active_arm].get().measured_joint_position
+        q = np.array([q.joint_1, q.joint_2, q.joint_3, q.joint_4, q.joint_5, q.joint_6, q.joint_7])
+        low = self.joint_lim_low - (q - self._joint_boundary)
+        low_violation = low[low > 0].sum()
+        high = (q + self._joint_boundary) - self.joint_lim_high
+        high_violation = high[high > 0].sum()
```

Comment:
Again if we used the move group abstraction getting joint angles would be a one liner

Reply:
How does the move group abstraction work?

Reply:
it's basically a set of joints with lots of relevant functionality. There's both python and C++ api for it, we use the python API in MoveitEnabledRobot